File: packages/guards/src/hallucinated-url.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

interface HallucinatedUrlOptions {
  action: 'block' | 'warn';
}

const URL_RE = /https?:\/\/[^\s"'<>)\]]+/gi;
const ARXIV_ID_RE = /^(\d{4}\.\d{4,5})(v\d+)?$/;

function checkUrl(url: string): string | null {
  // Internal/localhost URLs
  if (/\b(?:127\.0\.0\.1|localhost|\.local\b|\.internal\b)/i.test(url)) {
    return 'internal-url';
  }

  // Very long URLs
  if (url.length > 200) return 'excessively-long-url';

  // Fake arxiv
  const arxivMatch = url.match(/arxiv\.org\/abs\/(.+?)(?:[?#]|$)/);
  if (arxivMatch) {
    if (!ARXIV_ID_RE.test(arxivMatch[1])) return 'invalid-arxiv-id';
  }

  // Overly specific paths: 4+ segments of random-looking strings
  try {
    const parsed = new URL(url);
    const segments = parsed.pathname.split('/').filter((s) => s.length > 0);
    if (segments.length >= 4) {
      const randomSegments = segments.filter((s) => /^[a-f0-9]{8,}$/i.test(s) || /^[A-Za-z0-9_-]{20,}$/.test(s));
      if (randomSegments.length >= 2) return 'random-path-segments';
    }
  } catch {
    return 'malformed-url';
  }

  // Non-existent TLDs
  try {
    const host = new URL(url).hostname;
    const tld = host.split('.').pop() ?? '';
    const suspiciousTlds = ['xyz123', 'notreal', 'fake', 'test123'];
    if (suspiciousTlds.includes(tld.toLowerCase())) return 'suspicious-tld';
  } catch { /* already caught above */ }

  return null;
}

export function hallucinatedUrl(options: HallucinatedUrlOptions): Guard {
  return {
    name: 'hallucinated-url',
    version: '0.1.0',
    description: 'Detect fabricated/hallucinated URLs in LLM output',
    category: 'ai',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const urls = text.match(URL_RE) ?? [];
      const flagged: { url: string; reason: string }[] = [];
      for (const url of urls) {
        const reason = checkUrl(url);
        if (reason) flagged.push({ url, reason });
      }
      const triggered = flagged.length > 0;
      return {
        guardName: 'hallucinated-url',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        message: triggered
          ? `${flagged.length} potentially hallucinated URL(s) found`
          : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { flaggedUrls: flagged } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/hallucinated_url.py
"""Detect fabricated/hallucinated URLs in LLM output."""
from __future__ import annotations

import re
import time
from urllib.parse import urlparse

from open_guardrail.core import GuardResult

_URL_RE = re.compile(r"https?://[^\s\"'<>)\]]+", re.I)
_ARXIV_ID = re.compile(r"^(\d{4}\.\d{4,5})(v\d+)?$")


def _check(url: str) -> str | None:
    if re.search(r"\b(?:127\.0\.0\.1|localhost|\.local\b|\.internal\b)", url, re.I):
        return "internal-url"
    if len(url) > 200:
        return "excessively-long-url"
    arxiv = re.search(r"arxiv\.org/abs/(.+?)(?:[?#]|$)", url)
    if arxiv and not _ARXIV_ID.match(arxiv.group(1)):
        return "invalid-arxiv-id"
    try:
        parsed = urlparse(url)
        segs = [s for s in (parsed.path or "").split("/") if s]
        if len(segs) >= 4:
            rand = [s for s in segs if re.match(r"^[a-f0-9]{8,}$", s, re.I) or re.match(r"^[A-Za-z0-9_-]{20,}$", s)]
            if len(rand) >= 2:
                return "random-path-segments"
        tld = (parsed.hostname or "").split(".")[-1]
        if tld.lower() in ("xyz123", "notreal", "fake", "test123"):
            return "suspicious-tld"
    except Exception:
        return "malformed-url"
    return None


class _HallucinatedUrl:
    def __init__(self, *, action: str = "warn") -> None:
        self.name = "hallucinated-url"
        self.action = action

    def check(self, text: str, stage: str = "output") -> GuardResult:
        start = time.perf_counter()
        urls = _URL_RE.findall(text)
        flagged: list[dict[str, str]] = []
        for u in urls:
            reason = _check(u)
            if reason:
                flagged.append({"url": u, "reason": reason})
        triggered = len(flagged) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="hallucinated-url",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=f"{len(flagged)} potentially hallucinated URL(s)" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"flagged_urls": flagged} if triggered else None,
        )


def hallucinated_url(*, action: str = "warn") -> _HallucinatedUrl:
    return _HallucinatedUrl(action=action)
